pkg/controller/status/controller.go

```diff
@@ -0,0 +1,410 @@
+	defer c.lock.Unlock()
+	return c.reported.LastReportTime.Time
+}
+
+// SetLastReportedTime sets the last reported time in a thread-safe way.
+func (c *Controller) SetLastReportedTime(at time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.reported.LastReportTime.IsZero() {
+		klog.V(2).Infof("Initializing last reported time to %s", at.UTC().Format(time.RFC3339))
+	}
+	c.reported.LastReportTime.Time = at
+	c.triggerStatusUpdate()
+}
+
+// AddSources adds sources in a thread-safe way.
+// A source is used to monitor parts of the operator.
+func (c *Controller) AddSources(sources ...controllerstatus.Interface) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.sources = append(c.sources, sources...)
+}
+
+// Sources provides the sources in a thread-safe way.
+// A source is used to monitor parts of the operator.
+func (c *Controller) Sources() []controllerstatus.Interface {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.sources
+}
+
+func (c *Controller) merge(clusterOperator *configv1.ClusterOperator) *configv1.ClusterOperator {
+	// prime the object if it does not exist
+	if clusterOperator == nil {
+		clusterOperator = &configv1.ClusterOperator{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: c.name,
+			},
+		}
+	}
+
+	ctrlStatus := newControllerStatus()
+	// calculate the current controller state
+	allReady, lastTransition := c.currentControllerStatus(ctrlStatus)
+
+	clusterOperator = clusterOperator.DeepCopy()
+	now := time.Now()
+	if len(c.namespace) > 0 {
+		clusterOperator.Status.RelatedObjects = c.relatedObjects()
+	}
+
+	isInitializing := !allReady && now.Sub(c.controllerStartTime()) < 3*time.Minute
+
+	// cluster operator conditions
+	cs := newConditions(&clusterOperator.Status)
+	updateDisabledAndFailingConditions(cs, ctrlStatus, isInitializing, lastTransition)
+
+	// once the operator is running it is always considered available
+	cs.setCondition(configv1.OperatorAvailable, configv1.ConditionTrue, "AsExpected", "", metav1.Now())
+
+	updateProcessingConditionWithSummary(cs, ctrlStatus, isInitializing, lastTransition)
+
+	if release := os.Getenv("RELEASE_VERSION"); len(release) > 0 {
+		clusterOperator.Status.Versions = []configv1.OperandVersion{
+			{Name: "operator", Version: release},
+		}
+	}
+
+	reported := Reported{LastReportTime: metav1.Time{Time: c.LastReportedTime()}}
+	if data, err := json.Marshal(reported); err != nil {
+		klog.Errorf("Unable to marshal statusMessage extension: %v", err)
+	} else {
+		clusterOperator.Status.Extension.Raw = data
+	}
+	return clusterOperator
+}
+
+// calculate the current controller status based on its given sources
+func (c *Controller) currentControllerStatus(ctrlStatus *controllerStatus) (allReady bool, lastTransition time.Time) {
+	var errorReason string
+	var errs []string
+
+	allReady = false
+
+	for i, source := range c.Sources() {
+		summary, ready := source.CurrentStatus()
+		if !ready {
+			klog.V(4).Infof("Source %d %T is not ready", i, source)
+			allReady = true
```

Comment:
This doesn't seem right. If the source is not ready then `allReady = true` doesn't make sense IMO.

Reply:
Well it seems it's not working at the moment. From the CI:
>: operator conditions insights 
Operator unavailable (null): operator is not reporting conditions

I am also missing some unit tests for `conditions.go`, `controller.go`.

Reply:
You are right. Nice catch, thank you.